### ISSUE TITLE
mesh3d: Different lighting results with different hardware

### ISSUE THREAD

**Issue by jackparmer** [plotly/plotly.js#1368]
Consider this codepen:

http://codepen.io/monfera/full/ggByRW

On my 12 in Macbook, I see this:

![image](https://cloud.githubusercontent.com/assets/1865834/22805728/f5e07f68-eeec-11e6-8cfc-55b196a9915b.png)

@etpinard 's Thinkpad and @monfera 's Macbook show this:

![image](https://cloud.githubusercontent.com/assets/1865834/22805737/0535d7ce-eeed-11e6-946e-c0020f0b13be.png)

Another example - consider this codepen:

http://codepen.io/monfera/pen/XjxZjZ

Result on my 12 in Macbook:

![image](https://cloud.githubusercontent.com/assets/1865834/22805761/220d896e-eeed-11e6-8754-62c483b32a11.png)

Result on larger Macbooks and Thinkpad:

![image](https://cloud.githubusercontent.com/assets/1865834/22805787/3d66c130-eeed-11e6-9f1a-fad66913da2e.png)

***

I tried this in Chrome, Safari, and Firefox with the same result.

Computer specs:

![image](https://cloud.githubusercontent.com/assets/1865834/22805815/53162ab6-eeed-11e6-9ed2-ec1b417fdbb1.png)

Chrome specs:

![image](https://cloud.githubusercontent.com/assets/1865834/22805831/6378ab72-eeed-11e6-967b-bde5a64b9c47.png)

**Comment by rreusser** [comment]
Hmm… looks like flipped normals to me. I'd suggest maybe `gl_FrontFacing` isn't properly supported, but 2016 seems pretty recent for that.

http://stackoverflow.com/questions/24375171/is-there-a-reliable-alternative-to-gl-frontfacing-in-a-fragment-shader

**Comment by etpinard** [comment]
cc @dfcreative which have insight about this.

**Comment by monfera** [comment]
I don't have access to this machine but guessing is always fun:

- perhaps the `gl` floats are `lowp` - I think the standard is quite permissive about precision
- or we're using an extension that silently causes an issue

One thing to try: visiting a bunch of WebGL content on the web to see if *those* work properly.

**Comment by jackparmer** [comment]
Quick update: It doesn’t seem to matter what value I set for `diffuse` or `specular`. I can set them to 0.1 or 1 or 1000 and nothing changes.

**Comment by monfera** [comment]
Do these work fine? http://webglsamples.org/

**Comment by rreusser** [comment]
Gotchas like `pow(-abs(x), 2.0)` being inconsistent across platforms bit me a couple times, but I didn't happen to catch anything obvious with a quick glance.

**Comment by jackparmer** [comment]
> Do these work fine? http://webglsamples.org/

Yeah, nothing looks "off" with those. I can also change the `diffuse` and `specular` settings here:

http://multivis.net/lecture/phong.html

**Comment by monfera** [comment]
Not that it should inherently make it fail, but `mesh3d` uses the Cook-Torrance model. E.g. here one can choose the reflection model in a dropdown: http://www.calebjohnston.com/webgl.html

**Comment by jackparmer** [comment]
Seems to work too - I see this? 

![image](https://cloud.githubusercontent.com/assets/1865834/22806698/4b5aef10-eef1-11e6-9d10-d66866524167.png)

**Comment by rreusser** [comment]
Oops, sorry. I had the wrong module.

**Comment by monfera** [comment]
One more thing - possibly relating to the above possibility (lower precision floats): I needed to tweak the EPSILON value in the normals calculation. The brain mesh is made up of _very_ small triangles. With the wrong EPSILON value, whole swaths of lighting just don't happen. This is my current best guess. 

Way to test it: load some `mesh3d` content from eg.  `plot.ly` and see if _that_ works fine.

**Comment by jackparmer** [comment]
Hm, yeah, larger triangles _seem OK_:

https://plot.ly/~empet/13481/delaunay-triangulation-and-alpha-shapes-for-a-set-of-3d-points-1-delaunay-triang/
![image](https://cloud.githubusercontent.com/assets/1865834/22806986/c5e5a06c-eef2-11e6-8b83-afedfa716412.png)

https://plot.ly/~RPlotBot/3008/moebius-band-triangulation/
![image](https://cloud.githubusercontent.com/assets/1865834/22807005/da2260d8-eef2-11e6-870f-8fe9fc005773.png)

(I found these`mesh3d` plot.ly creations via: https://plot.ly/feed/?q=plottype:mesh3d)

**Comment by jackparmer** [comment]
Changing `diffuse` and `specular` still doesn't seem to have an effect though:

![image](https://cloud.githubusercontent.com/assets/1865834/22807182/a2a0a966-eef3-11e6-9242-3ff86dd5104f.png)

**Comment by monfera** [comment]
I added the epsilon attribs to the codepen, with their respective default values: http://codepen.io/monfera/pen/XjxZjZ

Could you test if lower values make it work? E.g. 1e-9 instead of 1e-6, or 1e-15 instead of 1e-12. Or some much more ambitious change.

**Comment by jackparmer** [comment]
Epsilon attribs don't seem to effect anything on my machine - I still get the dark and foggy brain no matter how I set them:

![image](https://cloud.githubusercontent.com/assets/1865834/22807323/4f6fb6a0-eef4-11e6-80d1-e485125ff41b.png)

**Comment by jackparmer** [comment]
Another example on a trisurf plot with less vertices:

https://plot.ly/~monfera/19

Robert sees this shiny helicopter:

![image](https://cloud.githubusercontent.com/assets/1865834/22807638/089e53d8-eef6-11e6-9582-bbfb0b652afa.png)

I see this midnight rider: 

![image](https://cloud.githubusercontent.com/assets/1865834/22807657/1e8d0df6-eef6-11e6-9e79-323eaa1990fe.png)

**Comment by jackparmer** [comment]
Here are some more brain images that might be useful in debugging. One shows specular reflection (is "shiny"). Another is an odd mix of shiny and unshiny.  

Shiny:

![image](https://cloud.githubusercontent.com/assets/1865834/22834088/0bc33420-ef82-11e6-8210-7ac81e44c302.png)

Odd mix of shiny + unshiny:

![image](https://cloud.githubusercontent.com/assets/1865834/22834108/15e4b474-ef82-11e6-86d6-ff929570d33e.png)

Dark:

![image](https://cloud.githubusercontent.com/assets/1865834/22834121/1c84c120-ef82-11e6-8ab9-bbde1dbd7ea1.png)

![image](https://cloud.githubusercontent.com/assets/1865834/22834128/22c808e4-ef82-11e6-8533-e7f5132b355f.png)


[Brains.zip](https://github.com/plotly/plotly.js/files/767170/Brains.zip)

**Comment by rreusser** [comment]
@jackparmer what does this look like for you?

https://rawgit.com/rreusser/057e5c31845192528cf62ba075540853/raw/index.html

It's just a simple sphere that uses some of the same basic lighting functions ([cook-torrance](https://github.com/stackgl/glsl-specular-cook-torrance) and [diffuse-lambert](https://github.com/stackgl/glsl-diffuse-lambert)). It looks like this for me:

<img width="423" alt="screen shot 2017-02-14 at 14 15 38" src="https://cloud.githubusercontent.com/assets/572717/22944855/186e7df6-f2c0-11e6-8b9a-928f3e18c25f.png">

Source [here](https://gist.githubusercontent.com/rreusser/057e5c31845192528cf62ba075540853/raw/83af93a9106e9d5190bc764953b0d734005e5258/test.js)

**Comment by jackparmer** [comment]
Looks a bit different, like this:

![image](https://cloud.githubusercontent.com/assets/1865834/22946127/8ae1d276-f2c4-11e6-87e9-4e2e435e3117.png)

**Comment by rreusser** [comment]
Ah, sorry. I rotated it in the code at the last minute. That's a 👍 from me on being the same. Not that I thought it would be different, but just a quick sanity check.

**Comment by dfcreative** [comment]
@jackparmer [Ricky](https://github.com/plotly/plotly.js/issues/1368#issuecomment-278799076) is right, `gl_FrontFacing` is broken [here](https://github.com/gl-vis/gl-mesh3d/blob/master/lib/triangle-fragment.glsl#L31-L33)

**Comment by etpinard** [comment]
From https://github.com/plotly/plotly.js/pull/1423#issuecomment-287411731 :

--------------------
After digging into the behavior quite a bit with @alexcjohnson, here's what we conclude:

The current behavior mostly make sense. The really weird stuff is all in the light positioning.

- the magnitudes have to be really giant. Are they pixels or something?
- z exhibits really weird behavior and thresholds

Apart from how it does work, here's how we think it should work:

The light should be positioned effectively in clip coordinates with the addition of an aspect ratio. Conceptual test cases:

- a light position of `[1, 0, 0]` should be just on the right side of the screen (*at the depth of the center*, that is. the projection matrix makes this statement otherwise meaningless). It involves an (inverse?) view matrix somewhere in there, I think.
- a light position of `[10, 0, 0]` would be roughly speaking 10 half-screens worth off to the right (again, at the depth of/relative to the center point with respect to the alignment of the eye)
- a light position of `[0, 1, 0]` should be at the top of the screen. If the plot is not square, then one or the other of these will need to take into account an aspect ratio.
- a light position of `[1, 1, 0]` should be at 45 degrees NE regardless of the plot's aspect ratio. This constraint probably determines where the aspect ratio goes
- a light position of `[0, 0, 1]` should be more or less *at* the eye position, unless I'm wrong about scales. At the very least in the direction of the eye. Out of the screen instead of into because of the righthand rule.
- a light position of `[0, 0, 10]` should be way behind the camera
- If we can, it'd be great of `length(light) > 1` implied the light tended not to fall *inside* the screen (I'm not 100% convinced this can be satisfied subject to the other constraints)


Note: scene aspect ratio and axis ranges seem to scale things entirely independently of the view, so I think they can be ignored. Open to feedback, but I think these constraints are enough to minimally fix the parts of the current behavior that are surprising without changing anything substantial. In talking with @alexcjohnson one conclusion is that existing behavior like the weird 1.00002 -> 1.00003 z threshold is best just fixed so that it's probably best *not* to preserve strict backward compatibility.
---------------

**Comment by jackparmer** [comment]
Hey @dfcreative I just ran into this issue again. Did https://github.com/gl-vis/gl-mesh3d/pull/8/ fix the dark `mesh3d` shading issue? I can bring in the laptop where this shows up if helpful. 
![image](https://user-images.githubusercontent.com/1865834/35650653-db9a9cb6-06aa-11e8-95c2-a493a5a2abce.png)

**Comment by jackparmer** [comment]
One things that is weird, I don't see any shading issues in this gl-mesh3d examples: 
https://gl-vis.github.io/gl-isosurface3d/brainbrowser.html
I would expect to see the same issue as the plotly.js `mesh3d` brains.
![image](https://user-images.githubusercontent.com/1865834/35650986-1297a744-06ac-11e8-8e52-de808e0a2ccf.png)

**Comment by monfera** [comment]
Just a minor note, these last two images aren't too different; both seem to be primarily or exclusively lit by ambient light; it appears that the darker one simply has a lower light level. A high ambient level is usually not good because it "washes away" detail and reduces plasticity, as can be seen on the brainbrowser example. Neither seem to have eg. the specular component, compare to [this](https://codepen.io/monfera/pen/XjxZjZ?q=brain%20pial&order=popularity&depth=everything&show_forks=true):

![image](https://user-images.githubusercontent.com/1548516/35652223-019c4454-06e3-11e8-8d7a-58c5be088a36.png)

In short, I think the dark brain above wants to be like what's seen in the codepen example, but misses components other than the (justly) subdued ambient component. I go by the look so I may well be wrong.

**Comment by jackparmer** [comment]
@monfera The issue is that these mesh's look completely different depending on the viewer's computer. Here's another example: https://plot.ly/~empet/14759/. This mesh look completely different on my laptop versus my phone. These are the exact same plot (no lighting as been adjusted).

This is a huge bug. Plotly meshes shouldn't look completely different depending on what computer you're using. 

iPhone:
![image](https://user-images.githubusercontent.com/1865834/35770381-0d5a80a6-08e8-11e8-95ef-0ce7fb6ffb53.png)

Laptop:
![image](https://user-images.githubusercontent.com/1865834/35770385-19cd03fe-08e8-11e8-868c-2e0783badf81.png)

**Comment by dfcreative** [comment]
We have that fixed in gl-mesh3d https://github.com/gl-vis/gl-mesh3d/pull/8, it takes a separate effort meeting the requirements here though

**Comment by jackparmer** [comment]
AFAICT it's a GPU issue. The MBP where Plotly.js meshes appear dark has an Intel 515 GPU, but the exact same MBP (same model, OS, and Chrome) with an Intel 5300 GPU displays meshes with "normal" lighting. 

The Plotly Cloud image server also renders Plotly.js meshes overly dark, wonder what GPU it's using ATM:

![image](https://user-images.githubusercontent.com/1865834/35772254-b204c902-0908-11e8-9c6f-295a539870dc.png)

**Comment by jackparmer** [comment]
Note: @nicolaskruchten also sees this with an _Intel Iris_ GPU on his MBP

**Comment by nicolaskruchten** [comment]
Here are my specs, for the record

![image](https://user-images.githubusercontent.com/203523/35829925-78739516-0a92-11e8-9287-2dbf51a73e83.png)

**Comment by nicolaskruchten** [comment]
FWIW the first few links now look great on my machine :)

**Comment by nicolaskruchten** [comment]
(ah, but they look fine with 1.35.0 also... What's an updated test case I can try?)

**Comment by etpinard** [comment]
> FWIW the first few links now look great on my machine :)

Strange. I haven't updated the CDN links yet.

**Comment by nicolaskruchten** [comment]
Sorry, too quick on the draw! I had initially seen these issues in a Dash app that showed a brain. Not sure what an easy-to-access test case is. Happy to check on anything pre/post if you like on my machine.

**Comment by etpinard** [comment]
https://cdn.plot.ly/plotly-1.36.0.min.js is now up!
